File: skellig-test-step-processing-rmq/src/main/java/org/skellig/teststep/processor/rmq/model/RmqDetails.java
package org.skellig.teststep.processor.rmq.model;


public class RmqDetails {

    private String channelId;
    private RmqHostDetails hostDetails;
    private RmqExchangeDetails exchange;
    private RmqQueueDetails queue;

    private RmqDetails(String channelId, RmqHostDetails hostDetails, RmqExchangeDetails exchange, RmqQueueDetails queue) {
        this.channelId = channelId;
        this.hostDetails = hostDetails;
        this.exchange = exchange;
        this.queue = queue;
    }

    public String getChannelId() {
        return channelId;
    }

    public RmqHostDetails getHostDetails() {
        return hostDetails;
    }

    public RmqExchangeDetails getExchange() {
        return exchange;
    }

    public RmqQueueDetails getQueue() {
        return queue;
    }

    public static class Builder {

        private String channelId;
        private RmqHostDetails hostDetails;
        private RmqExchangeDetails exchange;
        private RmqQueueDetails queue;

        public Builder withChannelId(String channelId) {
            this.channelId = channelId;
            return this;
        }

        public Builder withHostDetails(RmqHostDetails hostDetails) {
            this.hostDetails = hostDetails;
            return this;
        }

        public Builder withExchange(RmqExchangeDetails exchange) {
            this.exchange = exchange;
            return this;
        }

        public Builder withQueue(RmqQueueDetails queue) {
            this.queue = queue;
            return this;
        }

        public RmqDetails build() {
            return new RmqDetails(channelId, hostDetails, exchange, queue);
        }
    }
}

File: skellig-test-step-processing-rmq/src/main/kotlin/org/skellig/teststep/processor/rmq/model/RmqDetails.kt
package org.skellig.teststep.processor.rmq.model


/**
 * Represents the details of an RMQ connection, including the host details, exchange details, and queue details.
 *
 * @property hostDetails The details of the RMQ host.
 * @property exchange The details of the RMQ exchange.
 * @property queue The details of the RMQ queue.
 */
class RmqDetails private constructor(val hostDetails: RmqHostDetails,
                                     val exchange: RmqExchangeDetails,
                                     val queue: RmqQueueDetails) {

    override fun toString(): String {
        return "$hostDetails:$exchange:$queue"
    }

    class Builder {
        private var hostDetails: RmqHostDetails? = null
        private var exchange: RmqExchangeDetails? = null

        private var queue: RmqQueueDetails? = null

        fun hostDetails(hostDetails: RmqHostDetails?) = apply {
            this.hostDetails = hostDetails
        }

        fun exchange(exchange: RmqExchangeDetails?) = apply {
            this.exchange = exchange
        }

        fun queue(queue: RmqQueueDetails?) = apply {
            this.queue = queue
        }

        fun build(): RmqDetails {
            return RmqDetails(hostDetails!!, exchange!!, queue!!)
        }
    }
}